File: Solutions/0146-lru-cache/solution.cpp
struct Node {
    Node* next;
    Node* prev;
    int val; 
    int key;
};

class LRUCache {
private:
    Node* head = nullptr;
    Node* tail = nullptr;
    unordered_map<int, Node*> mp;
    int cap;
public:
    LRUCache(int capacity) {
        this->cap = capacity;
    }
    
    int get(int key) {
        if(mp.find(key) == mp.end()) return -1;
        Node* curr = mp[key];
        if(curr == tail) return curr->val;
        // what if curr is head?!
        if(curr->prev != nullptr) curr->prev->next = curr->next;
        if(curr->next != nullptr) curr->next->prev = curr->prev;
        if(curr == head) head = head ->next;
        tail->next = curr;
        curr->prev = tail;
        tail = curr;
        return curr->val;
    }
    
    void put(int key, int value) {
        // do we have to move when update?!
        if(get(key) != -1) {
            mp[key]->val = value;
            return;
        }
        Node* curr = new Node();
        mp[key] = curr;
        curr->val = value;
        curr->key = key;
        if(head == nullptr) {
            head = curr;
            tail = curr;
            return;
        }
        tail->next = curr;
        curr->prev = tail;
        tail = curr;
        if(mp.size() > cap) {
            // evict!
            mp.erase(head->key);
            head = head->next;
            head->prev = nullptr;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: Solutions/0146-lru-cache/solution.js
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.head = null;
    this.tail = null;
    this.mp = new Map();
    this.cap = capacity;
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if(this.mp.has(key) === false) return -1;
    const curr = this.mp.get(key);
    if(curr === this.tail) return curr.val;
    if(curr.prev !== null) curr.prev.next = curr.next;
    if(curr.next !== null) curr.next.prev = curr.prev;
    if(curr === this.head) this.head = this.head.next;
    this.tail.next = curr;
    curr.prev = this.tail;
    this.tail = curr;
    return curr.val;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if(this.get(key) !== -1) {
        this.mp.get(key).val = value;
        return;
    }
    const curr = {
        val: value,
        key: key,
        next: null,
        prev: null
    };
    this.mp.set(key, curr);
    if(this.head === null) {
        this.head = curr;
        this.tail = curr;
        return;
    }
    this.tail.next = curr;
    curr.prev = this.tail;
    this.tail = curr;
    if(this.mp.size > this.cap) {
        this.mp.delete(this.head.key);
        this.head = this.head.next;
        this.head.prev = null;
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
